classes/security/authorization/internal/UserAccessibleWorkflowStageRequiredPolicy.inc.php

```diff
@@ -78,30 +94,25 @@ function effect() {
 	 */
 	function _getAccessibleStageRoles($userId, $contextId, &$submission, $stageId) {
 		$stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO'); /* @var $stageAssignmentDao StageAssignmentDAO */
-		$userRoles = $this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES);
+		$stageAssignmentsResult = $stageAssignmentDao->getBySubmissionAndUserIdAndStageId($submission->getId(), $userId, $stageId);
 
 		$accessibleStageRoles = array();
-		foreach ($userRoles as $roleId) {
-			switch ($roleId) {
-				case ROLE_ID_MANAGER:
-					// Context managers have access to all submission stages.
-					$accessibleStageRoles[] = $roleId;
-					break;
-
-				case ROLE_ID_ASSISTANT:
-				case ROLE_ID_SUB_EDITOR:
-				case ROLE_ID_AUTHOR:
-					// The requested workflow stage has been assigned to them
-					// in the requested submission.
-					$stageAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($submission->getId(), $roleId, $stageId, $userId);
-					if(!$stageAssignments->wasEmpty()) {
-						$accessibleStageRoles[] = $roleId;
-					}
-					break;
-				default:
-					break;
+
+		// If unassigned, only managers and admins have access
+		if ($stageAssignmentsResult->wasEmpty()) {
+			$userRoles = $this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES);
+			$accessibleStageRoles = array_intersect(array(ROLE_ID_MANAGER, ROLE_ID_SITE_ADMIN), $userRoles);
+
+		// Assigned users have access based on their assignment
+		} else {
+			$userGroupDao = DAORegistry::getDAO('UserGroupDAO');
+			while ($stageAssignment = $stageAssignmentsResult->next()) {
+				$userGroup = $userGroupDao->getById($stageAssignment->getUserGroupId(), $contextId);
+				$accessibleStageRoles[] = $userGroup->getRoleId();
```

Comment:
What if that role Id is not one of (ROLE_ID_ASSISTANT, ROLE_ID_SUB_EDITOR, ROLE_ID_AUTHOR)? Earlier, in the removed switch statement only those roles were considered...

Reply:
Hmm, that's a good question! I don't believe that other user roles can be given a `StageAssignment` in a submission, can they? And if they can, then they probably should be granted access through this policy. I think?

Reply:
hmmm... yes, probably... as far as I can see, additionally to those roles only subscription manager can also be assigned in submission stage... so yes, maybe that role should then also have an access...

Reply:
:+1: